jwk/helper.go

```diff
@@ -44,12 +44,14 @@ func EnsureAsymmetricKeypairExists(ctx context.Context, r InternalRegistry, alg,
 }
 
 func GetOrGenerateKeys(ctx context.Context, r InternalRegistry, m Manager, set, kid, alg string) (private *jose.JSONWebKey, err error) {
-	getLock(set).Lock()
-	defer getLock(set).Unlock()
-
 	keys, err := m.GetKeySet(ctx, set)
+
 	if errors.Is(err, x.ErrNotFound) || keys != nil && len(keys.Keys) == 0 {
 		r.Logger().Warnf("JSON Web Key Set \"%s\" does not exist yet, generating new key pair...", set)
+
+		l := getLock(set)
+		defer l.Lock()
```

Comment:
In case you missed it, I think you meant to lock the returned lock then defer unlock. Same thing below on line 69.

At the end of this `if` block it may make sense to directly return `FindPrivateKey(keys)`. If we don't it opens up the potential for deadlock if `FindPrivateKey` fails and we try to generate another keyset. Also I don't really think it would help to generate another keyset if the one we just generated didn't have a private key.

Reply:
whoops!